app_dart/lib/src/request_handlers/refresh_chromebot_status.dart

```diff
@@ -79,19 +80,82 @@ class RefreshChromebotStatus extends ApiRequestHandler<Body> {
       DatastoreService datastore,
       Map<LuciBuilder, List<LuciTask>> luciTasks) async {
     for (String branch in branches) {
-      await for (FullTask task in datastore.queryRecentTasks(
-          taskName: builder.taskName, branch: branch)) {
-        for (LuciTask luciTask in luciTasks[builder]) {
-          if (luciTask.commitSha == task.commit.sha &&
-              luciTask.ref == 'refs/heads/$branch') {
+      final List<FullTask> tasks = await datastore
+          .queryRecentTasks(taskName: builder.taskName, branch: branch)
+          .toList();
+      final List<LuciTask> allLuciTasks = luciTasks[builder];
+      final Set<LuciTask> updatedLuciTasks = <LuciTask>{};
+
+      /// Scan both [tasks] and [luciTasks] from old to new, since [luciTasks] may
+      /// contain new re-run builds which are not in datastore yet. If matched, then
+      /// update accordingly.
+      for (int i = tasks.length - 1; i >= 0; i--) {
+        final FullTask task = tasks[i];
+        for (int j = allLuciTasks.length - 1; j >= 0; j--) {
+          final LuciTask luciTask = allLuciTasks[j];
+          if (_taskMatched(task, luciTask, branch) &&
+              !updatedLuciTasks.contains(luciTask)) {
             final Task update = task.task;
             update.status = luciTask.status;
+            update.buildId = luciTask.buildId;
             await datastore.insert(<Task>[update]);
-            // Stop updating task whenever we find the latest status of the same commit.
+            updatedLuciTasks.add(luciTask);
             break;
           }
         }
       }
+
+      final List<Task> newTasks =
+          _getNewTasks(updatedLuciTasks, builder, allLuciTasks, datastore);
+      await datastore.insert(newTasks);
+    }
+  }
+
+  bool _taskMatched(FullTask task, LuciTask luciTask, String branch) {
+    return luciTask.commitSha == task.commit.sha &&
+        luciTask.ref == 'refs/heads/$branch' &&
+        (task.task.buildId == null || task.task.buildId == luciTask.buildId);
+  }
+
+  /// Get new re-run luci builds by excluding existing updated ones. /api/
+  /// refresh-github-commits always inserts luci tasks for any new commit.
+  /// However when people re-run luci builds afterwards, there are no
+  /// corresponding tasks in datastore to be updated.
+  ///
+  /// When there are re-run luci builds, `_getNewTasks` returns them to be
+  /// inserted into datastore.
+  ///
+  /// The best time to add re-run luci builds to datastore is when
+  /// re-running them (push rather than pull).
+  // TODO(keyonghan): add new tasks to datastore when re-running luci builds,
+  // https://github.com/flutter/flutter/issues/58268
+  List<Task> _getNewTasks(Set<LuciTask> updatedLuciTasks, LuciBuilder builder,
+      List<LuciTask> allLuciTasks, DatastoreService datastore) {
+    final List<Task> tasks = <Task>[];
+    for (LuciTask luciTask in allLuciTasks) {
+      if (!updatedLuciTasks.contains(luciTask) &&
```

Comment:
Instead of filtering here why not just get the list of task we are interested in? both buildBucket and datastore supports retrieving only what is necessary.

Reply:
Have removed the section of `new builds insert` from this PR.